File: go/memory.go
package klyro

import (
	"context"
	"encoding/binary"
	"fmt"
	"math"
	"strconv"
	"strings"

	"github.com/redis/go-redis/v9"
)

type Mode string

const (
	Search Mode = "SEARCH"
	Vector Mode = "VECTOR"
	Hybrid Mode = "HYBRID"
)

type Metric string

const (
	Cosine       Metric = "COSINE"
	L2           Metric = "L2"
	InnerProduct Metric = "IP"
)

type Fusion string

const (
	Linear Fusion = "LINEAR"
	RRF    Fusion = "RRF"
)

type Condition string

const (
	NX Condition = "NX"
	XX Condition = "XX"
)

type Weights struct{ Keyword, Vector, Recency, Importance float64 }
type Filter struct {
	Field, Op string
	Value     any
}

type CreateOptions struct {
	Mode     Mode
	Dim      int
	Metric   Metric
	Weights  *Weights
	HalfLife int64
}
type ConfigOptions struct {
	Weights  *Weights
	HalfLife int64
}
type AddOptions struct {
	Text       string
	ID         string
	Vector     []float32
	Metadata   map[string]any
	Importance *float64
	TTL        int64
	Condition  Condition
}
type ReturnOptions struct{ NoText, WithMetadata, WithVector bool }
type SearchOptions struct {
	ReturnOptions
	TopK       int
	Filters    []Filter
	WithScores bool
}
type QueryOptions struct {
	SearchOptions
	Text    string
	Vector  []float32
	Weights *Weights
	Fusion  Fusion
}
type ScanOptions struct {
	Count   int
	Filters []Filter
}

type MemoryInfo struct {
	Mode                    Mode
	Dim                     int
	Metric                  Metric
	Weights                 Weights
	HalfLife                int64
	Records, Vectors, Terms int64
	AverageDocumentLength   float64
	Bytes                   int64
}
type MemoryRecord struct {
	ID                         string
	Text                       *string
	Importance                 float64
	CreatedAt, UpdatedAt, PTTL int64
	Metadata                   map[string]string
	Vector                     []float32
}
type MemoryHit struct {
	MemoryRecord
	Score                                   float64
	KeywordScore, VectorScore, RecencyScore *float64
}
type ScanResult struct {
	Cursor string
	IDs    []string
}

type MemoryClient struct{ redis redis.UniversalClient }

func (m *MemoryClient) call(ctx context.Context, command string, args ...any) (any, error) {
	all := append([]any{"MEM." + command}, args...)
	return m.redis.Do(ctx, all...).Result()
}

func (m *MemoryClient) Create(ctx context.Context, key string, o CreateOptions) error {
	mode := o.Mode
	if mode == "" {
		mode = Hybrid
	}
	metric := o.Metric
	if metric == "" {
		metric = Cosine
	}
	args := []any{key, "MODE", string(mode)}
	if o.Dim > 0 {
		args = append(args, "DIM", o.Dim)
	}
	args = append(args, "METRIC", string(metric))
	args = appendWeights(args, o.Weights)
	if o.HalfLife > 0 {
		args = append(args, "HALFLIFE", o.HalfLife)
	}
	_, err := m.call(ctx, "CREATE", args...)
	return err
}

func (m *MemoryClient) Info(ctx context.Context, key string) (MemoryInfo, error) {
	raw, err := m.call(ctx, "INFO", key)
	if err != nil {
		return MemoryInfo{}, err
	}
	v, err := mapping(raw)
	if err != nil {
		return MemoryInfo{}, err
	}
	w, err := mapping(v["weights"])
	if err != nil {
		return MemoryInfo{}, err
	}
	return MemoryInfo{Mode: Mode(text(v["mode"])), Dim: integer(v["dim"]), Metric: Metric(text(v["metric"])),
		Weights:  Weights{number(w["keyword"]), number(w["vector"]), number(w["recency"]), number(w["importance"])},
		HalfLife: int64(integer(v["halflife"])), Records: int64(integer(v["records"])), Vectors: int64(integer(v["vectors"])),
		Terms: int64(integer(v["terms"])), AverageDocumentLength: number(v["avg_doc_len"]), Bytes: int64(integer(v["bytes"]))}, nil
}

func (m *MemoryClient) Config(ctx context.Context, key string, o ConfigOptions) error {
	args := appendWeights([]any{key}, o.Weights)
	if o.HalfLife > 0 {
		args = append(args, "HALFLIFE", o.HalfLife)
	}
	if len(args) == 1 {
		return fmt.Errorf("klyro: weights or half-life is required")
	}
	_, err := m.call(ctx, "CONFIG", args...)
	return err
}
func (m *MemoryClient) Card(ctx context.Context, key string) (int64, error) {
	raw, err := m.call(ctx, "CARD", key)
	return int64(integer(raw)), err
}

func (m *MemoryClient) Add(ctx context.Context, key string, o AddOptions) (string, error) {
	args := []any{key, "TEXT", o.Text}
	if o.ID != "" {
		args = append(args, "ID", o.ID)
	}
	if o.Vector != nil {
		args = append(args, "VEC", encodeVector(o.Vector))
	}
	for k, v := range o.Metadata {
		args = append(args, "META", k, v)
	}
	if o.Importance != nil {
		args = append(args, "IMPORTANCE", *o.Importance)
	}
	if o.TTL > 0 {
		args = append(args, "TTL", o.TTL)
	}
	if o.Condition != "" {
		args = append(args, string(o.Condition))
	}
	raw, err := m.call(ctx, "ADD", args...)
	return text(raw), err
}

func (m *MemoryClient) Get(ctx context.Context, key, id string, o ReturnOptions) (*MemoryRecord, error) {
	args := appendReturns([]any{key, id}, o, false)
	raw, err := m.call(ctx, "GET", args...)
	if err != nil || raw == nil {
		return nil, err
	}
	record, _, err := decodeRecord(raw)
	return &record, err
}
func (m *MemoryClient) MGet(ctx context.Context, key string, ids ...string) ([]*MemoryRecord, error) {
	if len(ids) == 0 {
		return nil, fmt.Errorf("klyro: at least one id is required")
	}
	args := []any{key}
	for _, id := range ids {
		args = append(args, id)
	}
	raw, err := m.call(ctx, "MGET", args...)
	if err != nil {
		return nil, err
	}
	items, err := sequence(raw)
	if err != nil {
		return nil, err
	}
	out := make([]*MemoryRecord, len(items))
	for i, item := range items {
		if item != nil {
			rec, _, e := decodeRecord(item)
			if e != nil {
				return nil, e
			}
			out[i] = &rec
		}
	}
	return out, nil
}
func (m *MemoryClient) Delete(ctx context.Context, key string, ids ...string) (int64, error) {
	return m.countCall(ctx, "DEL", key, ids)
}
func (m *MemoryClient) SetMetadata(ctx context.Context, key, id string, values map[string]any) (int64, error) {
	if len(values) == 0 {
		return 0, fmt.Errorf("klyro: metadata must not be empty")
	}
	args := []any{key, id}
	for k, v := range values {
		args = append(args, k, v)
	}
	raw, err := m.call(ctx, "SETMETA", args...)
	return int64(integer(raw)), err
}
func (m *MemoryClient) DeleteMetadata(ctx context.Context, key, id string, fields ...string) (int64, error) {
	if len(fields) == 0 {
		return 0, fmt.Errorf("klyro: at least one field is required")
	}
	args := []any{key, id}
	for _, f := range fields {
		args = append(args, f)
	}
	raw, err := m.call(ctx, "DELMETA", args...)
	return int64(integer(raw)), err
}
func (m *MemoryClient) Expire(ctx context.Context, key, id string, seconds int64) (bool, error) {
	raw, err := m.call(ctx, "EXPIRE", key, id, seconds)
	return integer(raw) == 1, err
}

func (m *MemoryClient) Scan(ctx context.Context, key, cursor string, o ScanOptions) (ScanResult, error) {
	args := []any{key, cursor}
	if o.Count > 0 {
		args = append(args, "COUNT", o.Count)
	}
	args = appendFilters(args, o.Filters)
	raw, err := m.call(ctx, "SCAN", args...)
	if err != nil {
		return ScanResult{}, err
	}
	pair, err := sequence(raw)
	if err != nil || len(pair) != 2 {
		return ScanResult{}, fmt.Errorf("klyro: invalid scan reply")
	}
	ids, err := sequence(pair[1])
	if err != nil {
		return ScanResult{}, err
	}
	out := ScanResult{Cursor: text(pair[0]), IDs: make([]string, len(ids))}
	for i := range ids {
		out.IDs[i] = text(ids[i])
	}
	return out, nil
}
func (m *MemoryClient) Search(ctx context.Context, key, query string, o SearchOptions) ([]MemoryHit, error) {
	return m.hits(ctx, "SEARCH", appendRetrieval([]any{key, query}, o)...)
}
func (m *MemoryClient) VectorSearch(ctx context.Context, key string, vector []float32, o SearchOptions) ([]MemoryHit, error) {
	return m.hits(ctx, "VSEARCH", appendRetrieval([]any{key, "VEC", encodeVector(vector)}, o)...)
}
func (m *MemoryClient) Query(ctx context.Context, key string, o QueryOptions) ([]MemoryHit, error) {
	if o.Text == "" && o.Vector == nil {
		return nil, fmt.Errorf("klyro: text or vector is required")
	}
	args := []any{key}
	if o.Text != "" {
		args = append(args, "TEXT", o.Text)
	}
	if o.Vector != nil {
		args = append(args, "VEC", encodeVector(o.Vector))
	}
	if o.Fusion != "" {
		args = append(args, "FUSION", string(o.Fusion))
	}
	args = appendWeights(args, o.Weights)
	args = appendRetrieval(args, o.SearchOptions)
	return m.hits(ctx, "QUERY", args...)
}

func (m *MemoryClient) hits(ctx context.Context, command string, args ...any) ([]MemoryHit, error) {
	raw, err := m.call(ctx, command, args...)
	if err != nil {
		return nil, err
	}
	items, err := sequence(raw)
	if err != nil {
		return nil, err
	}
	out := make([]MemoryHit, len(items))
	for i, item := range items {
		_, hit, e := decodeRecord(item)
		if e != nil {
			return nil, e
		}
		out[i] = hit
	}
	return out, nil
}
func (m *MemoryClient) countCall(ctx context.Context, command, key string, values []string) (int64, error) {
	if len(values) == 0 {
		return 0, fmt.Errorf("klyro: at least one item is required")
	}
	args := []any{key}
	for _, v := range values {
		args = append(args, v)
	}
	raw, err := m.call(ctx, command, args...)
	return int64(integer(raw)), err
}

func appendWeights(args []any, w *Weights) []any {
	if w != nil {
		args = append(args, "WEIGHTS", w.Keyword, w.Vector, w.Recency, w.Importance)
	}
	return args
}
func appendFilters(args []any, filters []Filter) []any {
	for _, f := range filters {
		args = append(args, "FILTER", f.Field, strings.ToUpper(f.Op), f.Value)
	}
	return args
}
func appendReturns(args []any, o ReturnOptions, scores bool) []any {
	if o.NoText {
		args = append(args, "NOTEXT")
	}
	if o.WithMetadata {
		args = append(args, "WITHMETA")
	}
	if o.WithVector {
		args = append(args, "WITHVEC")
	}
	if scores {
		args = append(args, "WITHSCORES")
	}
	return args
}
func appendRetrieval(args []any, o SearchOptions) []any {
	if o.TopK > 0 {
		args = append(args, "TOPK", o.TopK)
	}
	args = appendFilters(args, o.Filters)
	return appendReturns(args, o.ReturnOptions, o.WithScores)
}
func encodeVector(values []float32) []byte {
	out := make([]byte, len(values)*4)
	for i, v := range values {
		binary.LittleEndian.PutUint32(out[i*4:], math.Float32bits(v))
	}
	return out
}
func decodeVector(raw any) ([]float32, error) {
	var b []byte
	switch value := raw.(type) {
	case []byte:
		b = value
	case string:
		b = []byte(value)
	default:
		return nil, fmt.Errorf("klyro: invalid vector reply")
	}
	if len(b)%4 != 0 {
		return nil, fmt.Errorf("klyro: invalid vector reply")
	}
	out := make([]float32, len(b)/4)
	for i := range out {
		out[i] = math.Float32frombits(binary.LittleEndian.Uint32(b[i*4:]))
	}
	return out, nil
}
func text(v any) string {
	switch x := v.(type) {
	case string:
		return x
	case []byte:
		return string(x)
	default:
		return fmt.Sprint(x)
	}
}
func number(v any) float64 {
	switch x := v.(type) {
	case float64:
		return x
	case int64:
		return float64(x)
	default:
		n, _ := strconv.ParseFloat(text(x), 64)
		return n
	}
}
func integer(v any) int {
	switch x := v.(type) {
	case int:
		return x
	case int64:
		return int(x)
	default:
		n, _ := strconv.Atoi(text(x))
		return n
	}
}
func sequence(v any) ([]any, error) {
	if x, ok := v.([]any); ok {
		return x, nil
	}
	return nil, fmt.Errorf("klyro: invalid array reply")
}
func mapping(v any) (map[string]any, error) {
	if x, ok := v.(map[any]any); ok {
		out := map[string]any{}
		for k, val := range x {
			out[text(k)] = val
		}
		return out, nil
	}
	if x, ok := v.(map[string]any); ok {
		return x, nil
	}
	seq, err := sequence(v)
	if err != nil || len(seq)%2 != 0 {
		return nil, fmt.Errorf("klyro: invalid map reply")
	}
	out := map[string]any{}
	for i := 0; i < len(seq); i += 2 {
		out[text(seq[i])] = seq[i+1]
	}
	return out, nil
}
func decodeRecord(raw any) (MemoryRecord, MemoryHit, error) {
	v, err := mapping(raw)
	if err != nil {
		return MemoryRecord{}, MemoryHit{}, err
	}
	r := MemoryRecord{ID: text(v["id"]), Importance: number(v["importance"]), CreatedAt: int64(integer(v["created_at"])), UpdatedAt: int64(integer(v["updated_at"])), PTTL: int64(integer(v["pttl"]))}
	if x, ok := v["text"]; ok {
		s := text(x)
		r.Text = &s
	}
	if x, ok := v["meta"]; ok {
		values, e := mapping(x)
		if e != nil {
			return r, MemoryHit{}, e
		}
		r.Metadata = map[string]string{}
		for k, value := range values {
			r.Metadata[k] = text(value)
		}
	}
	if x, ok := v["vector"]; ok && x != nil {
		r.Vector, err = decodeVector(x)
		if err != nil {
			return r, MemoryHit{}, err
		}
	}
	h := MemoryHit{MemoryRecord: r, Score: number(v["score"])}
	for name, target := range map[string]**float64{"keyword_score": &h.KeywordScore, "vector_score": &h.VectorScore, "recency_score": &h.RecencyScore} {
		if x, ok := v[name]; ok {
			n := number(x)
			*target = &n
		}
	}
	return r, h, nil
}

File: python/src/klyro_db/memory.py
from __future__ import annotations

from dataclasses import dataclass, field
from struct import pack, unpack
from typing import Literal, Mapping, Sequence, TypeAlias, cast

from redis import Redis

BytesLike: TypeAlias = str | bytes
Vector: TypeAlias = Sequence[float] | bytes
Mode: TypeAlias = Literal["SEARCH", "VECTOR", "HYBRID"]
Metric: TypeAlias = Literal["COSINE", "L2", "IP"]
Fusion: TypeAlias = Literal["LINEAR", "RRF"]
FilterOp: TypeAlias = Literal["EQ", "NE", "GT", "GTE", "LT", "LTE", "IN", "CONTAINS"]
Condition: TypeAlias = Literal["NX", "XX"]


@dataclass(frozen=True)
class Weights:
    keyword: float = 0.35
    vector: float = 0.50
    recency: float = 0.10
    importance: float = 0.05


@dataclass(frozen=True)
class Filter:
    field: BytesLike
    op: FilterOp
    value: BytesLike | int | float


@dataclass(frozen=True)
class MemoryCreate:
    mode: Mode = "HYBRID"
    dim: int | None = None
    metric: Metric = "COSINE"
    weights: Weights | None = None
    halflife: int | None = None


@dataclass(frozen=True)
class MemoryConfig:
    weights: Weights | None = None
    halflife: int | None = None


@dataclass(frozen=True)
class MemoryAdd:
    text: BytesLike
    id: BytesLike | None = None
    vector: Vector | None = None
    metadata: Mapping[BytesLike, BytesLike | int | float] = field(default_factory=dict)
    importance: float | None = None
    ttl: int | None = None
    condition: Condition | None = None


@dataclass(frozen=True)
class MemoryReturn:
    no_text: bool = False
    with_metadata: bool = False
    with_vector: bool = False


@dataclass(frozen=True)
class MemorySearch(MemoryReturn):
    top_k: int | None = None
    filters: Sequence[Filter] = ()
    with_scores: bool = False


@dataclass(frozen=True)
class MemoryQuery(MemorySearch):
    text: BytesLike | None = None
    vector: Vector | None = None
    weights: Weights | None = None
    fusion: Fusion | None = None


@dataclass(frozen=True)
class MemoryScan:
    count: int | None = None
    filters: Sequence[Filter] = ()


@dataclass(frozen=True)
class MemoryInfo:
    mode: Mode
    dim: int
    metric: Metric
    weights: Weights
    halflife: int
    records: int
    vectors: int
    terms: int
    avg_doc_len: float
    bytes: int


@dataclass(frozen=True)
class MemoryRecord:
    id: str
    text: str | None
    importance: float
    created_at: int
    updated_at: int
    pttl: int
    metadata: dict[str, str] | None = None
    vector: tuple[float, ...] | None = None


@dataclass(frozen=True)
class MemoryHit(MemoryRecord):
    score: float = 0.0
    keyword_score: float | None = None
    vector_score: float | None = None
    recency_score: float | None = None


@dataclass(frozen=True)
class MemoryScanResult:
    cursor: str
    ids: list[str]


def _text(value: object) -> str:
    return value.decode() if isinstance(value, bytes) else str(value)


def _mapping(value: object) -> dict[str, object]:
    if isinstance(value, dict):
        return {_text(k): v for k, v in value.items()}
    values = cast(Sequence[object], value)
    if len(values) % 2:
        raise TypeError("invalid memory map reply")
    return {_text(values[i]): values[i + 1] for i in range(0, len(values), 2)}


def _vector(value: Vector) -> bytes:
    if isinstance(value, bytes):
        return value
    values = [float(item) for item in value]
    return pack(f"<{len(values)}f", *values)


def _decode_vector(value: object) -> tuple[float, ...] | None:
    if value is None:
        return None
    raw = cast(bytes, value)
    if len(raw) % 4:
        raise TypeError("invalid vector reply")
    return unpack(f"<{len(raw) // 4}f", raw)


def _weights(args: list[object], value: Weights | None) -> None:
    if value is not None:
        args.extend(("WEIGHTS", value.keyword, value.vector, value.recency, value.importance))


def _filters(args: list[object], values: Sequence[Filter]) -> None:
    for item in values:
        args.extend(("FILTER", item.field, item.op, item.value))


def _returns(args: list[object], options: MemoryReturn) -> None:
    if options.no_text: args.append("NOTEXT")
    if options.with_metadata: args.append("WITHMETA")
    if options.with_vector: args.append("WITHVEC")
    if isinstance(options, MemorySearch) and options.with_scores: args.append("WITHSCORES")


def _retrieval(args: list[object], options: MemorySearch) -> None:
    if options.top_k is not None: args.extend(("TOPK", options.top_k))
    _filters(args, options.filters)
    _returns(args, options)


class MemoryClient:
    """Typed access to every currently implemented MEM.* command."""

    def __init__(self, redis: Redis) -> None:
        self._redis = redis

    def _call(self, command: str, *args: object) -> object:
        return self._redis.execute_command(f"MEM.{command}", *args)  # type: ignore[no-untyped-call]

    def _record(self, reply: object) -> MemoryRecord | MemoryHit | None:
        if reply is None:
            return None
        data = _mapping(reply)
        record_id = _text(data["id"])
        record_text = _text(data["text"]) if "text" in data else None
        importance = float(cast(float, data["importance"]))
        created_at = int(cast(int, data["created_at"]))
        updated_at = int(cast(int, data["updated_at"]))
        pttl = int(cast(int, data["pttl"]))
        metadata = {_text(k): _text(v) for k, v in _mapping(data["meta"]).items()} if "meta" in data else None
        vector = _decode_vector(data["vector"]) if "vector" in data else None
        if "score" not in data:
            return MemoryRecord(record_id, record_text, importance, created_at, updated_at, pttl, metadata, vector)
        def component(name: str) -> float | None:
            return float(cast(float, data[name])) if name in data else None
        return MemoryHit(record_id, record_text, importance, created_at, updated_at, pttl, metadata, vector, float(cast(float, data["score"])), component("keyword_score"), component("vector_score"), component("recency_score"))

    def create(self, key: BytesLike, options: MemoryCreate) -> Literal["OK"]:
        args: list[object] = [key, "MODE", options.mode]
        if options.dim is not None: args.extend(("DIM", options.dim))
        args.extend(("METRIC", options.metric)); _weights(args, options.weights)
        if options.halflife is not None: args.extend(("HALFLIFE", options.halflife))
        return cast(Literal["OK"], _text(self._call("CREATE", *args)))

    def info(self, key: BytesLike) -> MemoryInfo:
        data = _mapping(self._call("INFO", key)); w = _mapping(data["weights"])
        return MemoryInfo(cast(Mode, _text(data["mode"])), int(cast(int, data["dim"])), cast(Metric, _text(data["metric"])), Weights(*(float(cast(float, w[k])) for k in ("keyword", "vector", "recency", "importance"))), int(cast(int, data["halflife"])), int(cast(int, data["records"])), int(cast(int, data["vectors"])), int(cast(int, data["terms"])), float(cast(float, data["avg_doc_len"])), int(cast(int, data["bytes"])))

    def config(self, key: BytesLike, options: MemoryConfig) -> Literal["OK"]:
        args: list[object] = [key]; _weights(args, options.weights)
        if options.halflife is not None: args.extend(("HALFLIFE", options.halflife))
        if len(args) == 1: raise ValueError("weights or halflife is required")
        return cast(Literal["OK"], _text(self._call("CONFIG", *args)))

    def card(self, key: BytesLike) -> int: return int(cast(int, self._call("CARD", key)))

    def add(self, key: BytesLike, options: MemoryAdd) -> str:
        args: list[object] = [key, "TEXT", options.text]
        if options.id is not None: args.extend(("ID", options.id))
        if options.vector is not None: args.extend(("VEC", _vector(options.vector)))
        for name, value in options.metadata.items(): args.extend(("META", name, value))
        if options.importance is not None: args.extend(("IMPORTANCE", options.importance))
        if options.ttl is not None: args.extend(("TTL", options.ttl))
        if options.condition is not None: args.append(options.condition)
        return _text(self._call("ADD", *args))

    def get(self, key: BytesLike, id: BytesLike, options: MemoryReturn = MemoryReturn()) -> MemoryRecord | None:
        args: list[object] = [key, id]; _returns(args, options)
        result = self._record(self._call("GET", *args))
        return result if isinstance(result, MemoryRecord) else None

    def mget(self, key: BytesLike, id: BytesLike, *ids: BytesLike) -> list[MemoryRecord | None]:
        return [result if isinstance(result := self._record(item), MemoryRecord) else None for item in cast(Sequence[object], self._call("MGET", key, id, *ids))]

    def delete(self, key: BytesLike, id: BytesLike, *ids: BytesLike) -> int: return int(cast(int, self._call("DEL", key, id, *ids)))
    def set_metadata(self, key: BytesLike, id: BytesLike, metadata: Mapping[BytesLike, BytesLike | int | float]) -> int:
        if not metadata: raise ValueError("metadata must not be empty")
        args: list[object] = [key, id]
        for name, value in metadata.items(): args.extend((name, value))
        return int(cast(int, self._call("SETMETA", *args)))
    def delete_metadata(self, key: BytesLike, id: BytesLike, field: BytesLike, *fields: BytesLike) -> int: return int(cast(int, self._call("DELMETA", key, id, field, *fields)))
    def expire(self, key: BytesLike, id: BytesLike, seconds: int) -> bool: return bool(self._call("EXPIRE", key, id, seconds))

    def scan(self, key: BytesLike, cursor: str | int, options: MemoryScan = MemoryScan()) -> MemoryScanResult:
        args: list[object] = [key, cursor]
        if options.count is not None: args.extend(("COUNT", options.count))
        _filters(args, options.filters); reply = cast(Sequence[object], self._call("SCAN", *args))
        return MemoryScanResult(_text(reply[0]), [_text(item) for item in cast(Sequence[object], reply[1])])

    def search(self, key: BytesLike, query: BytesLike, options: MemorySearch = MemorySearch()) -> list[MemoryHit]:
        args: list[object] = [key, query]; _retrieval(args, options)
        return [cast(MemoryHit, self._record(item)) for item in cast(Sequence[object], self._call("SEARCH", *args))]

    def vector_search(self, key: BytesLike, vector: Vector, options: MemorySearch = MemorySearch()) -> list[MemoryHit]:
        args: list[object] = [key, "VEC", _vector(vector)]; _retrieval(args, options)
        return [cast(MemoryHit, self._record(item)) for item in cast(Sequence[object], self._call("VSEARCH", *args))]

    def query(self, key: BytesLike, options: MemoryQuery) -> list[MemoryHit]:
        if options.text is None and options.vector is None: raise ValueError("text or vector is required")
        args: list[object] = [key]
        if options.text is not None: args.extend(("TEXT", options.text))
        if options.vector is not None: args.extend(("VEC", _vector(options.vector)))
        if options.fusion is not None: args.extend(("FUSION", options.fusion))
        _weights(args, options.weights); _retrieval(args, options)
        return [cast(MemoryHit, self._record(item)) for item in cast(Sequence[object], self._call("QUERY", *args))]
